plugins/shuffle/round.py

```diff
@@ -227,41 +239,45 @@ def process_shuffling(self):
         4. If player is last player he goes to the next phase ("Broadcasts Output") and broadcast the outputs.
         """
         phase = self.messages.phases[self.phase]
-        accused = None
+        accused, sender = None, None
         if self.me == self.last_player():
             sender = self.players[self.previous_player(player=self.last_player())]
-            if self.inbox[phase].get(sender):
-                self.messages.packets.ParseFromString(self.inbox[phase][sender])
-                try:
-                    for packet in self.messages.packets.packet:
-                        packet.packet.message.str = self.crypto.decrypt(packet.packet.message.str)
-                    self.messages.add_str(self.addr_new)
-                    self.messages.shuffle_packets()
-                    self.phase = 'BroadcastOutput'
-                    self.send_message()
-                    self.log_message("encrypt new address")
-                except CryptoError:
-                    accused = sender
-        else:
-            sender = self.players[self.previous_player()]
-            if self.inbox[phase].get(sender):
-                self.messages.packets.ParseFromString(self.inbox[phase][sender])
-                try:
-                    for packet in self.messages.packets.packet:
-                        packet.packet.message.str = self.crypto.decrypt(packet.packet.message.str)
-                    if not self.different_ciphertexts():
-                        self.messages.add_str(self.encrypt_new_address())
+            msg = self.inbox[phase].get(sender)
+            if msg:
+                accused = self._parse_from_string_safe(msg, sender)
```

Comment:
result if the message parses correctly: `accused` = `None`
result if the message fails: `accused` == `sender`

This logic is hard to follow but I think it is correct in this context. The usage of the function is kind of inverted from the way it is used in the announcement method.

Reply:
Nevermind. I see that this is the way it's really intended to be used. the `"bad"` name in the other method was a little confusing.

Reply:
I agree it's hard to follow .. but I had to write a function that fit reasonably within the confines of this half-assed code.

:/